File: typescript/src/bittorrent/krpc.ts
// SPDX-License-Identifier: MIT
// KRPC (BEP-5) DHT messages over bencode.
import { decode, encode, type BValue } from './bencode.js';

export function encodeQuery(tx: Buffer, method: string, args: { [k: string]: BValue }): Buffer {
  return encode({ t: tx, y: Buffer.from('q', 'latin1'), q: Buffer.from(method, 'utf8'), a: args });
}

export function encodeResponse(tx: Buffer, response: { [k: string]: BValue }): Buffer {
  return encode({ t: tx, y: Buffer.from('r', 'latin1'), r: response });
}

export function encodeError(tx: Buffer, code: number, message: string): Buffer {
  return encode({ t: tx, y: Buffer.from('e', 'latin1'), e: [code, Buffer.from(message, 'utf8')] });
}

export interface KrpcDecoded {
  transactionId: Buffer;
  type: string;
  method?: string;
  arguments?: { [k: string]: BValue };
  response?: { [k: string]: BValue };
  errorCode?: number;
  errorMessage?: string;
}

export function decodeKrpc(data: Buffer): KrpcDecoded {
  const d = decode(data) as { [k: string]: BValue };
  const y = (d['y'] as Buffer).toString('latin1');
  const out: KrpcDecoded = { transactionId: d['t'] as Buffer, type: y };
  if (y === 'q') {
    out.method = (d['q'] as Buffer).toString('utf8');
    out.arguments = d['a'] as { [k: string]: BValue };
  } else if (y === 'r') {
    out.response = d['r'] as { [k: string]: BValue };
  } else if (y === 'e') {
    const e = d['e'] as BValue[];
    if (e.length >= 2) {
      out.errorCode = e[0] as number;
      out.errorMessage = (e[1] as Buffer).toString('utf8');
    }
  }
  return out;
}

File: rust/src/bittorrent/krpc.rs
// SPDX-License-Identifier: MIT
//! KRPC (BEP-5) DHT messages over bencode.

use super::bencode::{self, Bencode};
use std::collections::BTreeMap;

pub fn encode_query(tx: &[u8], method: &str, args: BTreeMap<Vec<u8>, Bencode>) -> Vec<u8> {
    let mut d = BTreeMap::new();
    d.insert(b"t".to_vec(), Bencode::Bytes(tx.to_vec()));
    d.insert(b"y".to_vec(), Bencode::Bytes(b"q".to_vec()));
    d.insert(b"q".to_vec(), Bencode::Bytes(method.as_bytes().to_vec()));
    d.insert(b"a".to_vec(), Bencode::Dict(args));
    bencode::encode(&Bencode::Dict(d))
}

pub fn encode_response(tx: &[u8], response: BTreeMap<Vec<u8>, Bencode>) -> Vec<u8> {
    let mut d = BTreeMap::new();
    d.insert(b"t".to_vec(), Bencode::Bytes(tx.to_vec()));
    d.insert(b"y".to_vec(), Bencode::Bytes(b"r".to_vec()));
    d.insert(b"r".to_vec(), Bencode::Dict(response));
    bencode::encode(&Bencode::Dict(d))
}

pub fn encode_error(tx: &[u8], code: i64, message: &str) -> Vec<u8> {
    let mut d = BTreeMap::new();
    d.insert(b"t".to_vec(), Bencode::Bytes(tx.to_vec()));
    d.insert(b"y".to_vec(), Bencode::Bytes(b"e".to_vec()));
    d.insert(
        b"e".to_vec(),
        Bencode::List(vec![Bencode::Int(code), Bencode::Bytes(message.as_bytes().to_vec())]),
    );
    bencode::encode(&Bencode::Dict(d))
}
